flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/DataGenerators.java

```diff
@@ -104,6 +104,8 @@ public void cancel() {
 		if (secureProps != null) {
 			props.putAll(testServer.getSecureProperties());
 		}
+		// Ensure the producer enables idempotence.
+		props.putAll(testServer.getIdempotentProducerConfig());
```

Comment:
Is this the only place where we should be using `enable.idempotence`? 

Do I remember correctly that `enable.idempotence` is automatically/effectively set to true when using transactional kafka producer?

Reply:
Yes, you are right. The idempotence will be enabled if `transactional.id` is set. But our KafkaITCase is using AT_LEAST_ONCE semantic without enabling checkpoint, so by default a non-transactional producer will be instantiated, which will remove the `transactional.id` config. 
https://github.com/apache/flink/blob/bd56224c3063fd23d508a4250e5698d4840fa488/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java#L1155